initialization/initialization.go

```diff
@@ -275,17 +300,22 @@ func (init *Initializer) initFile(computeProviderID uint, fileIndex int, numLabe
 	currentPosition := numLabelsWritten
 	outputChan := make(chan []byte, 1024)
 
-	errGroup, ctx := errgroup.WithContext(context.Background())
+	errGroup, ctx := errgroup.WithContext(ctx)
 
 	// Start compute worker.
 	errGroup.Go(func() error {
 		defer close(outputChan)
 
 		for currentPosition < fileNumLabels {
 			select {
-			case <-init.stopChan:
+			case <-ctx.Done():
 				init.logger.Info("initialization: stopped")
-				return ErrStopped
+
+				if res := gpu.Stop(); res != gpu.StopResultOk {
```

Comment:
I did some digging, `gpu.Stop()` calls this C function with a `ms_timeout` value of 2000: https://github.com/spacemeshos/gpu-post/blob/develop/src/api_internal.cpp#L108

It essentially sets `g_spacemesh_api_abort_flag` to `true` which is used as a signal to terminate ongoing work like e.g. here: https://github.com/spacemeshos/gpu-post/blob/develop/src/cuda/driver-cuda.cpp#L220 (and in a few other places)

If initialization is not interrupted but continues to completion `gpu.Stop()` does not need to be called. A better name for that function would probably be `Abort()` or `Cancel()`

PS.: looking through this code I believe the whole thing is inherently racy. `g_spacemesh_api_abort_flag` is a global variable that is written to / read from different go routines in our application, unless the `volatile` keyword makes that OK I think there is a mutex missing that protects access to `g_spacemesh_api_abort_flag`.

Reply:
😮 The Go wrapper could provide a safe abstraction over the C library. If the C library `gpu-post` has a global state and thus can only be safely called once at the same time (no concurrency), then, IMHO, the Go wrapper around it could prevent calling it concurrently. It could, for example, have a global mutex locked on "start" and unlocked on "stop". Or at least panic if "start" is called twice (IMO panic is better then UB).

Otherwise, it should be at least documented in `gpu/api.go`.

Reply:
I created an issue to fix the racy `Stop()` call in a future PR here: https://github.com/spacemeshos/gpu-post/issues/79